File: server/src/main/java/com/uiptv/util/StalkerPortalParser.java
package com.uiptv.util;

import com.uiptv.model.Account;
import com.uiptv.service.AccountService;

import java.util.*;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.stream.Collectors;

import static com.uiptv.util.StringUtils.isNotBlank;
import static com.uiptv.util.UiptUtils.getNameFromUrl;

/**
 * Handles parsing of Stalker Portal accounts.
 */
public class StalkerPortalParser implements AccountParser {
    private static final List<StalkerAttributeParser> attributeParsers = Arrays.asList(
            new UrlAttributeParser(),
            new MacAttributeParser(),
            new SerialCutAttributeParser(),
            new SerialAttributeParser(),
            new SignatureAttributeParser(),
            new DeviceId1AttributeParser(),
            new DeviceId2AttributeParser()
    );

    private final Function<String, Account> accountProvider;
    private final Consumer<Account> accountSaver;

    public StalkerPortalParser() {
        this(
            name -> AccountService.getInstance().getByName(name),
            account -> AccountService.getInstance().save(account)
        );
    }

    public StalkerPortalParser(Function<String, Account> accountProvider, Consumer<Account> accountSaver) {
        this.accountProvider = accountProvider;
        this.accountSaver = accountSaver;
    }

    @Override
    public List<Account> parseAndSave(String text, boolean groupAccountsByMac, boolean convertM3uToXtreme) {
        String sanitizedText = sanitizeInput(text);
        List<Account> parsedAccounts = new ArrayList<>();
        Account account = new Account();
        MacAttributeParser macParser = new MacAttributeParser();
        UrlAttributeParser urlParser = new UrlAttributeParser();
        String lastSeenUrl = null;
        for (String line : sanitizedText.split("\\R")) {
            String trimmedLine = line.trim();
            if (trimmedLine.isEmpty()) continue;

            ParsedLine parsedLine = parseLine(line, trimmedLine, urlParser, macParser);
            lastSeenUrl = parsedLine.url != null ? parsedLine.url : lastSeenUrl;
            if (shouldStartNewAccount(account, parsedLine.mac)) {
                addParsedAccount(parsedAccounts, account);
                account = newAccountWithLastSeenUrl(lastSeenUrl);
            }
            applyLineMetadata(account, trimmedLine);
            applyParsedAttributes(account, line);
        }
        addParsedAccount(parsedAccounts, account);
        return saveAccounts(parsedAccounts, groupAccountsByMac);
    }

    private void applyValueToAccount(Account account, String value, StalkerAttributeType type) {
        switch (type) {
            case URL:
                if (!isNotBlank(account.getUrl())) {
                    account.setUrl(value);
                    account.setType(AccountType.STALKER_PORTAL);
                }
                break;
            case MAC:
                account.setMacAddress(value);
                break;
            case SERIAL:
                if (!isNotBlank(account.getSerialNumber())) {
                    account.setSerialNumber(value);
                }
                break;
            case SERIAL_CUT:
                account.setSerialNumber(value);
                break;
            case DEVICE_ID_1:
                account.setDeviceId1(value);
                break;
            case DEVICE_ID_2:
                account.setDeviceId2(value);
                break;
            case SIGNATURE:
                if (!isNotBlank(account.getSignature())) {
                    account.setSignature(value);
                }
                break;
        }
    }

    private List<Account> saveAccounts(List<Account> accounts, boolean groupAccountsByMac) {
        Map<String, Account> groupedAccounts = new LinkedHashMap<>();
        List<Account> individualAccounts = new ArrayList<>();
        List<Account> createdAccounts = new ArrayList<>();
        Set<String> processedNames = new HashSet<>();

        List<Account> validAccounts = accounts.stream()
                .filter(acc -> isNotBlank(acc.getUrl()) && isNotBlank(acc.getMacAddress()))
                .toList();

        for (Account currentAccount : validAccounts) {
            if (groupAccountsByMac) {
                saveGroupedAccount(currentAccount, groupedAccounts, createdAccounts, processedNames);
            } else {
                saveIndividualAccount(currentAccount, individualAccounts, createdAccounts, processedNames);
            }
        }

        individualAccounts.forEach(accountSaver);
        groupedAccounts.values().forEach(accountSaver);
        return createdAccounts;
    }

    private String sanitizeInput(String text) {
        return Arrays.stream(text.split("\\R"))
                .filter(line -> !(line.contains("http") && line.contains("?") && line.contains("=")))
                .map(UiptUtils::sanitizeStalkerText)
                .collect(Collectors.joining("\n"));
    }

    private ParsedLine parseLine(String line, String trimmedLine, UrlAttributeParser urlParser, MacAttributeParser macParser) {
        String url = urlParser.parse(line);
        String mac = url == null ? macParser.parse(line) : null;
        return new ParsedLine(trimmedLine, url, mac);
    }

    private boolean shouldStartNewAccount(Account account, String mac) {
        return mac != null && isNotBlank(account.getMacAddress());
    }

    private void addParsedAccount(List<Account> parsedAccounts, Account account) {
        if (isNotBlank(account.getUrl()) && isNotBlank(account.getMacAddress())) {
            parsedAccounts.add(account);
        }
    }

    private Account newAccountWithLastSeenUrl(String lastSeenUrl) {
        Account account = new Account();
        if (lastSeenUrl != null) {
            account.setUrl(lastSeenUrl);
            account.setType(AccountType.STALKER_PORTAL);
        }
        return account;
    }

    private void applyLineMetadata(Account account, String trimmedLine) {
        if ("POST".equalsIgnoreCase(trimmedLine)) {
            account.setHttpMethod("POST");
        }
        String detectedTimezone = detectTimezone(trimmedLine);
        if (detectedTimezone != null) {
            account.setTimezone(detectedTimezone);
        }
    }

    private void applyParsedAttributes(Account account, String line) {
        for (StalkerAttributeParser parser : attributeParsers) {
            String value = parser.parse(line);
            if (value != null) {
                StalkerAttributeType type = parser.getAttributeType();
                applyValueToAccount(account, value, type);
                // If a combined device id (e.g., "1/2") is provided, apply the same value to device id 2 as well
                if (type == StalkerAttributeType.DEVICE_ID_1) {
                    String lower = line.toLowerCase();
                    if (lower.matches(".*1\\s*/\\s*2.*")) {
                        applyValueToAccount(account, value, StalkerAttributeType.DEVICE_ID_2);
                    }
                }
                return;
            }
        }
    }


    private void saveGroupedAccount(Account currentAccount, Map<String, Account> groupedAccounts,
                                    List<Account> createdAccounts, Set<String> processedNames) {
        String name = getNameFromUrl(currentAccount.getUrl().replace("_", ""));
        currentAccount.setAccountName(name);
        Account existing = groupedAccounts.get(name);
        if (existing != null) {
            appendMacAddress(existing, currentAccount.getMacAddress());
            mergeExtraParams(existing, currentAccount);
            return;
        }
        Account existingInDb = accountProvider.apply(name);
        if (existingInDb != null) {
            appendMacAddress(existingInDb, currentAccount.getMacAddress());
            mergeExtraParams(existingInDb, currentAccount);
            groupedAccounts.put(name, existingInDb);
            processedNames.add(existingInDb.getAccountName());
            return;
        }
        currentAccount.setMacAddressList(currentAccount.getMacAddress());
        groupedAccounts.put(name, currentAccount);
        processedNames.add(name);
        createdAccounts.add(currentAccount);
    }

    private void saveIndividualAccount(Account currentAccount, List<Account> individualAccounts,
                                       List<Account> createdAccounts, Set<String> processedNames) {
        String uniqueName = nextUniqueAccountName(currentAccount.getUrl(), processedNames);
        currentAccount.setAccountName(uniqueName);
        currentAccount.setMacAddressList(currentAccount.getMacAddress());
        individualAccounts.add(currentAccount);
        processedNames.add(uniqueName);
        createdAccounts.add(currentAccount);
    }

    private String nextUniqueAccountName(String url, Set<String> processedNames) {
        String baseName = getNameFromUrl(url);
        int counter = 1;
        String uniqueName = baseName + "(" + counter + ")";
        while (accountProvider.apply(uniqueName) != null || processedNames.contains(uniqueName)) {
            counter++;
            uniqueName = baseName + "(" + counter + ")";
        }
        return uniqueName;
    }

    private void appendMacAddress(Account account, String macAddress) {
        String macList = account.getMacAddressList() != null ? account.getMacAddressList() : account.getMacAddress();
        if (macList == null || macList.isBlank()) {
            account.setMacAddressList(macAddress);
            return;
        }
        if (!macList.contains(macAddress)) {
            account.setMacAddressList(macList + "," + macAddress);
        }
    }

    private void mergeExtraParams(Account target, Account source) {
        if (!isNotBlank(target.getSerialNumber()) && isNotBlank(source.getSerialNumber())) {
            target.setSerialNumber(source.getSerialNumber());
        }
        if (!isNotBlank(target.getDeviceId1()) && isNotBlank(source.getDeviceId1())) {
            target.setDeviceId1(source.getDeviceId1());
        }
        if (!isNotBlank(target.getDeviceId2()) && isNotBlank(source.getDeviceId2())) {
            target.setDeviceId2(source.getDeviceId2());
        }
        if (!isNotBlank(target.getSignature()) && isNotBlank(source.getSignature())) {
            target.setSignature(source.getSignature());
        }
        if (!isNotBlank(target.getTimezone()) && isNotBlank(source.getTimezone())) {
            target.setTimezone(source.getTimezone());
        }
        if (!isNotBlank(target.getHttpMethod()) && isNotBlank(source.getHttpMethod())) {
            target.setHttpMethod(source.getHttpMethod());
        }
    }

    private record ParsedLine(String trimmedLine, String url, String mac) {
    }

    /**
     * Detects if the given line contains a valid timezone identifier.
     * Performs case-insensitive matching against all available ZoneIds.
     *
     * @param line The line to check for timezone
     * @return The matched timezone string, or null if no timezone is found
     */
    private String detectTimezone(String line) {
        if (line == null || line.trim().isEmpty()) {
            return null;
        }

        String lowerLine = line.toLowerCase();

        // Get all available zone IDs and perform case-insensitive contains matching
        for (String zoneId : java.time.ZoneId.getAvailableZoneIds()) {
            if (lowerLine.contains(zoneId.toLowerCase())) {
                return zoneId;
            }
        }

        return null;
    }
}

File: server/src/main/kotlin/com/uiptv/util/StalkerPortalParser.kt
package com.uiptv.util

import com.uiptv.api.AccountParser
import com.uiptv.api.StalkerAttributeParser
import com.uiptv.model.Account
import com.uiptv.service.AccountService
import java.time.ZoneId
import java.util.LinkedHashMap
import java.util.Locale
import java.util.function.Consumer
import java.util.function.Function

class StalkerPortalParser(
    private val accountProvider: Function<String, Account?> = Function { AccountService.getInstance().getByName(it) },
    private val accountSaver: Consumer<Account> = Consumer { AccountService.getInstance().save(it) }
) : AccountParser {
    override fun parseAndSave(text: String, groupAccountsByMac: Boolean, convertM3uToXtreme: Boolean): List<Account> {
        val sanitizedText = sanitizeInput(text)
        val parsedAccounts = ArrayList<Account>()
        var account = Account()
        val macParser = MacAttributeParser()
        val urlParser = UrlAttributeParser()
        var lastSeenUrl: String? = null

        for (line in sanitizedText.split(Regex("\\R"))) {
            val trimmedLine = line.trim()
            if (trimmedLine.isEmpty()) {
                continue
            }

            val parsedLine = parseLine(line, trimmedLine, urlParser, macParser)
            lastSeenUrl = parsedLine.url ?: lastSeenUrl
            if (shouldStartNewAccount(account, parsedLine.mac)) {
                addParsedAccount(parsedAccounts, account)
                account = newAccountWithLastSeenUrl(lastSeenUrl)
            }
            applyLineMetadata(account, trimmedLine)
            applyParsedAttributes(account, line)
        }
        addParsedAccount(parsedAccounts, account)
        return saveAccounts(parsedAccounts, groupAccountsByMac)
    }

    private fun applyValueToAccount(account: Account, value: String, type: StalkerAttributeType) {
        when (type) {
            StalkerAttributeType.URL -> if (!StringUtils.isNotBlank(account.url)) {
                account.url = value
                account.type = AccountType.STALKER_PORTAL
            }
            StalkerAttributeType.MAC -> account.macAddress = value
            StalkerAttributeType.SERIAL -> if (!StringUtils.isNotBlank(account.serialNumber)) {
                account.serialNumber = value
            }
            StalkerAttributeType.SERIAL_CUT -> account.serialNumber = value
            StalkerAttributeType.DEVICE_ID_1 -> account.deviceId1 = value
            StalkerAttributeType.DEVICE_ID_2 -> account.deviceId2 = value
            StalkerAttributeType.SIGNATURE -> if (!StringUtils.isNotBlank(account.signature)) {
                account.signature = value
            }
        }
    }

    private fun saveAccounts(accounts: List<Account>, groupAccountsByMac: Boolean): List<Account> {
        val groupedAccounts = LinkedHashMap<String, Account>()
        val groupedExtraAccounts = LinkedHashMap<String, Account>()
        val individualAccounts = ArrayList<Account>()
        val createdAccounts = ArrayList<Account>()
        val processedNames = HashSet<String>()

        val validAccounts = accounts.filter { StringUtils.isNotBlank(it.url) && StringUtils.isNotBlank(it.macAddress) }
        for (currentAccount in validAccounts) {
            if (groupAccountsByMac) {
                if (hasExtraParams(currentAccount)) {
                    saveGroupedExtraParamAccount(currentAccount, groupedExtraAccounts, individualAccounts, createdAccounts, processedNames)
                } else {
                    saveGroupedAccount(currentAccount, groupedAccounts, createdAccounts, processedNames)
                }
            } else {
                saveIndividualAccount(currentAccount, individualAccounts, createdAccounts, processedNames)
            }
        }

        individualAccounts.forEach(accountSaver::accept)
        groupedAccounts.values.forEach(accountSaver::accept)
        return createdAccounts
    }

    private fun sanitizeInput(text: String): String =
        text.split(Regex("\\R"))
            .filterNot { it.contains("http") && it.contains("?") && it.contains("=") }
            .joinToString("\n") { UiptUtils.sanitizeStalkerText(it) }

    private fun parseLine(
        line: String,
        trimmedLine: String,
        urlParser: UrlAttributeParser,
        macParser: MacAttributeParser
    ): ParsedLine {
        val url = urlParser.parse(line)
        val mac = if (url == null) macParser.parse(line) else null
        return ParsedLine(trimmedLine, url, mac)
    }

    private fun shouldStartNewAccount(account: Account, mac: String?): Boolean =
        mac != null && StringUtils.isNotBlank(account.macAddress)

    private fun addParsedAccount(parsedAccounts: MutableList<Account>, account: Account) {
        if (StringUtils.isNotBlank(account.url) && StringUtils.isNotBlank(account.macAddress)) {
            parsedAccounts += account
        }
    }

    private fun newAccountWithLastSeenUrl(lastSeenUrl: String?): Account {
        val account = Account()
        if (lastSeenUrl != null) {
            account.url = lastSeenUrl
            account.type = AccountType.STALKER_PORTAL
        }
        return account
    }

    private fun applyLineMetadata(account: Account, trimmedLine: String) {
        if ("POST".equals(trimmedLine, true)) {
            account.httpMethod = "POST"
        }
        detectTimezone(trimmedLine)?.let { account.timezone = it }
    }

    private fun applyParsedAttributes(account: Account, line: String) {
        for (parser in ATTRIBUTE_PARSERS) {
            val value = parser.parse(line) ?: continue
            val type = parser.getAttributeType()
            applyValueToAccount(account, value, type)
            if (type == StalkerAttributeType.DEVICE_ID_1) {
                val lower = line.lowercase()
                if (Regex(".*1\\s*/\\s*2.*").matches(lower)) {
                    applyValueToAccount(account, value, StalkerAttributeType.DEVICE_ID_2)
                }
            }
            return
        }
    }

    private fun hasExtraParams(account: Account): Boolean =
        StringUtils.isNotBlank(account.serialNumber) ||
            StringUtils.isNotBlank(account.deviceId1) ||
            StringUtils.isNotBlank(account.deviceId2) ||
            StringUtils.isNotBlank(account.signature)

    private fun saveGroupedAccount(
        currentAccount: Account,
        groupedAccounts: MutableMap<String, Account>,
        createdAccounts: MutableList<Account>,
        processedNames: MutableSet<String>
    ) {
        val name = UiptUtils.getNameFromUrl((currentAccount.url ?: "").replace("_", ""))
        currentAccount.accountName = name
        val existing = groupedAccounts[name]
        if (existing != null) {
            appendMacAddress(existing, currentAccount.macAddress ?: "")
            mergeExtraParams(existing, currentAccount)
            return
        }
        val existingInDb = accountProvider.apply(name)
        if (existingInDb != null) {
            appendMacAddress(existingInDb, currentAccount.macAddress ?: "")
            mergeExtraParams(existingInDb, currentAccount)
            groupedAccounts[name] = existingInDb
            existingInDb.accountName?.let(processedNames::add)
            return
        }
        currentAccount.macAddressList = currentAccount.macAddress
        groupedAccounts[name] = currentAccount
        processedNames.add(name)
        createdAccounts += currentAccount
    }

    private fun saveGroupedExtraParamAccount(
        currentAccount: Account,
        groupedExtraAccounts: MutableMap<String, Account>,
        individualAccounts: MutableList<Account>,
        createdAccounts: MutableList<Account>,
        processedNames: MutableSet<String>
    ) {
        val identityKey = buildExtraParamIdentityKey(currentAccount)
        val existing = groupedExtraAccounts[identityKey]
        if (existing != null) {
            appendMacAddress(existing, currentAccount.macAddress ?: "")
            mergeExtraParams(existing, currentAccount)
            return
        }

        val existingInDb = findExistingExtraParamAccount(currentAccount)
        if (existingInDb != null) {
            appendMacAddress(existingInDb, currentAccount.macAddress ?: "")
            mergeExtraParams(existingInDb, currentAccount)
            groupedExtraAccounts[identityKey] = existingInDb
            individualAccounts += existingInDb
            existingInDb.accountName?.let(processedNames::add)
            return
        }

        saveIndividualAccount(currentAccount, individualAccounts, createdAccounts, processedNames)
        groupedExtraAccounts[identityKey] = currentAccount
    }

    private fun saveIndividualAccount(
        currentAccount: Account,
        individualAccounts: MutableList<Account>,
        createdAccounts: MutableList<Account>,
        processedNames: MutableSet<String>
    ) {
        val uniqueName = nextUniqueAccountName(currentAccount.url ?: "", processedNames)
        currentAccount.accountName = uniqueName
        currentAccount.macAddressList = currentAccount.macAddress
        individualAccounts += currentAccount
        processedNames.add(uniqueName)
        createdAccounts += currentAccount
    }

    private fun nextUniqueAccountName(url: String, processedNames: Set<String>): String {
        val baseName = UiptUtils.getNameFromUrl(url)
        var counter = 1
        var uniqueName = "$baseName($counter)"
        while (accountProvider.apply(uniqueName) != null || processedNames.contains(uniqueName)) {
            counter++
            uniqueName = "$baseName($counter)"
        }
        return uniqueName
    }

    private fun appendMacAddress(account: Account, macAddress: String) {
        val macList = account.macAddressList ?: account.macAddress
        if (macList.isNullOrBlank()) {
            account.macAddressList = macAddress
            return
        }
        if (!macList.contains(macAddress)) {
            account.macAddressList = "$macList,$macAddress"
        }
    }

    private fun findExistingExtraParamAccount(account: Account): Account? {
        val baseName = UiptUtils.getNameFromUrl((account.url ?: "").replace("_", ""))
        val directMatch = accountProvider.apply(baseName)
        if (isSameExtraParamAccount(directMatch, account)) {
            return directMatch
        }
        for (counter in 1 until 1000) {
            val candidate = accountProvider.apply("$baseName($counter)") ?: return null
            if (isSameExtraParamAccount(candidate, account)) {
                return candidate
            }
        }
        return null
    }

    private fun isSameExtraParamAccount(candidate: Account?, account: Account): Boolean =
        candidate != null &&
            hasExtraParams(candidate) &&
            normalizeIdentityValue(UiptUtils.getNameFromUrl((candidate.url ?: "").replace("_", ""))) ==
            normalizeIdentityValue(UiptUtils.getNameFromUrl((account.url ?: "").replace("_", ""))) &&
            normalizeIdentityValue(candidate.serialNumber) == normalizeIdentityValue(account.serialNumber) &&
            normalizeIdentityValue(candidate.deviceId1) == normalizeIdentityValue(account.deviceId1) &&
            normalizeIdentityValue(candidate.deviceId2) == normalizeIdentityValue(account.deviceId2) &&
            normalizeIdentityValue(candidate.signature) == normalizeIdentityValue(account.signature)

    private fun buildExtraParamIdentityKey(account: Account): String =
        listOf(
            normalizeIdentityValue(UiptUtils.getNameFromUrl((account.url ?: "").replace("_", ""))),
            normalizeIdentityValue(account.serialNumber),
            normalizeIdentityValue(account.deviceId1),
            normalizeIdentityValue(account.deviceId2),
            normalizeIdentityValue(account.signature)
        ).joinToString("|")

    private fun normalizeIdentityValue(value: String?): String = value?.trim()?.lowercase(Locale.ROOT) ?: ""

    private fun mergeExtraParams(target: Account, source: Account) {
        if (!StringUtils.isNotBlank(target.serialNumber) && StringUtils.isNotBlank(source.serialNumber)) {
            target.serialNumber = source.serialNumber
        }
        if (!StringUtils.isNotBlank(target.deviceId1) && StringUtils.isNotBlank(source.deviceId1)) {
            target.deviceId1 = source.deviceId1
        }
        if (!StringUtils.isNotBlank(target.deviceId2) && StringUtils.isNotBlank(source.deviceId2)) {
            target.deviceId2 = source.deviceId2
        }
        if (!StringUtils.isNotBlank(target.signature) && StringUtils.isNotBlank(source.signature)) {
            target.signature = source.signature
        }
        if (!StringUtils.isNotBlank(target.timezone) && StringUtils.isNotBlank(source.timezone)) {
            target.timezone = source.timezone
        }
        if (!StringUtils.isNotBlank(target.httpMethod) && StringUtils.isNotBlank(source.httpMethod)) {
            target.httpMethod = source.httpMethod
        }
    }

    private fun detectTimezone(line: String?): String? {
        if (line.isNullOrBlank()) {
            return null
        }
        val lowerLine = line.lowercase()
        for (zoneId in ZoneId.getAvailableZoneIds()) {
            if (lowerLine.contains(zoneId.lowercase())) {
                return zoneId
            }
        }
        return null
    }

    private data class ParsedLine(val trimmedLine: String, val url: String?, val mac: String?)

    private companion object {
        val ATTRIBUTE_PARSERS: List<StalkerAttributeParser> = listOf(
            UrlAttributeParser(),
            MacAttributeParser(),
            SerialCutAttributeParser(),
            SerialAttributeParser(),
            SignatureAttributeParser(),
            DeviceId1AttributeParser(),
            DeviceId2AttributeParser()
        )
    }
}
